### ISSUE TITLE
Level 7. Save

### ISSUE THREAD

**Issue by OhJunMing** [nus-cs2103-AY2021S2/forum#63]
[Question requirements] Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up. You may hard-code the file name and location e.g., [project_root]/data/duke.txt

Hi all, how do we save the tasklist data, do we just 
1. Save once for a single duke session initiate>>manipulate>>terminate
E.g User initiate session, adds stuff but terminate prematurely, any task added would not be saved to hard disk.

2, In a single duke session, save multiple times for every manipulation[add, delete],

**Comment by damithc** [comment]
Good question @OhJunMing . Go with the _better_ option. That is, discuss and decide which option is better.

**Comment by tjtanjin** [comment]
From the point of a user, I believe the second option is more intuitive and friendly. I would probably appreciate it more if every successful operation is saved such that a random crash later on won't affect all previous operations xD

Edit: On a relevant note, the discussion can also be taken further to discuss how saving/loading is done. For every operation, do you write to the file and then read from it again to repopulate/update the array list of task that you have, or do you have another piece of code to modify the already loaded array list of task? 

I think with the current size of the program and a task limit of 100, this would not matter but scalability wise, if the number of tasks start growing significantly, then it might get expensive to keep reading/writing all tasks from the file.

**Comment by SoonKeatNeo** [comment]
Personally, I find that the second option would be good. What if the program crashes or encounters some other unrecoverable error before the save has happened? I think the efficiency of saving/deleting/modifying the file can be called into question as well when you have a larger number of Tasks.

Inadvertently, this does also raise the question of what happens if the user does the following "unsupported" task:
1. Save a Task
2. Delete the storage file
3. Try to save the task again

This will probably be classified as unsupported behavior and not an intended use of the application. However, should we as developers foresee these types of issues and attempt to make provisions for them?

**Comment by tjtanjin** [comment]
> 1. Save a Task
> 2. Delete the storage file
> 3. Try to save the task again

I think this would be handled as an IOException within the method responsible for saving tasks. Ideally the code should re-generate the missing folders/file if they are found to not exist but upon a failed creation (maybe due to permission issues), it should terminate the entire program with a proper exit message.

**Comment by SoonKeatNeo** [comment]
> I think this would be handled as an IOException within the method responsible for saving tasks. Ideally the code should re-generate the missing folders/file if they are found to not exist but upon a failed creation (probably due to permission issues), it should terminate the entire program with a proper exit message.

That's true. I was considering this possibility because the way I implemented my add task function only appends the new task to the end of the opened file. 

If I were to re-generate missing file structures, then I would also need to consider adding previously existing tasks in the list to the file as well, otherwise, there will be a mismatch between what is in the file vs what is in the task list at run-time. This could cause problems as well if the user modifies a task added before the file was re-generated (e.g. mark a task as done in the application, and this task is present in memory but not present in the file on the user's hard drive). 

And there could be various more "unsupported" uses of the application or ways to break it like this that I hadn't really thought about as well.

**Comment by tjtanjin** [comment]
> If I were to re-generate missing file structures, then I would also need to consider adding previously existing tasks in the list to the file as well, otherwise, there will be a mismatch between what is in the file vs what is in the task list at run-time. This could cause problems as well if the user modifies a task added before the file was re-generated (e.g. mark a task as done in the application, and this task is present in memory but not present in the file on the user's hard drive).

Now that you say this I didn't actually consider re-generating the file with previously existing tasks even though that sounds like a best case scenario for the user - more patch up to do :'DDD

**Comment by rajobasu** [comment]
There are two perspectives to this problem I feel: 
- From a user standpoint, saving every change is obviously optimal. I feel we should give the user a choice to save only when pressing save, or save automatically (as is the case in many applications). 
- Is it fast if there is a lot of changes? 

Well, naivest way to do is every time you load all the tasks in the memory, do wtvr the user asks you to do, and before exiting just save everything to the file. 
- this is bad becoz if user terminates by some other way, all changes are lost. 
- Good becoz it is fast since everything is done in memory

Second way is to do the same thing but read/write again and again everytime a single operation is done
- this is good bcoz robust
- bad becoz VERY slow and nonscalable. 

Next method is in the file, instead of reading/writing everything, save the commands themselves. This will allow you to not write everything, but just append to the end of the file the new commands. Again we can do this in two flavours, either add after every user operation, or add at the end. (or maybe add at certain time intervals — maybe by having a separate Thread for this instead of doing on the main process thread). 
- Robust, user doesn’t lose information
- Writing is fast since we are appending new information instead of rewriting everything.
- Reading is costly becoz we read all operations performed before instead of all tasks. This is not an issue now since per task we can do atmost 3 operations — insert, markAsDone and remove. so #operations = O(#tasks). But this might not be so always. 
- Reading the first time is also costly, but not prohibitively so.

**Comment by damithc** [comment]
Thanks for adding a detailed analysis @rajobasu 

> * bad becoz VERY slow and nonscalable.

This is a point that could benefit from further analysis. At what point the speed and scalability becomes a problem? Furthermore, will it ever become a problem in normal usage?

**Comment by richardcom** [comment]
> Thanks for adding a detailed analysis @rajobasu
> 
> > * bad becoz VERY slow and nonscalable.
> 
> This is a point that could benefit from further analysis. At what point the speed and scalability becomes a problem? Furthermore, will it ever become a problem in normal usage?

Thanks prof, I think that maybe as the ip is a task management tool without network connection. Thus, I think that each TODO/DDL/... item is about 40 - 50 bytes in size in the txt file. Thus if we store 1000 tasks in the system, the txt file would be about 50kb, and the response time normally would be less than 0.1 sec. So maybe it would be acceptable when we consider small scale normal usage in Non-Functional Requriement.

**Comment by damithc** [comment]
> Thanks prof, I think that maybe as the ip is a task management tool without network connection. Thus, I think that each TODO/DDL/... item is about 40 - 50 bytes in size in the txt file. Thus if we store 1000 tasks in the system, the txt file would be about 50kb, and the response time normally would be less than 0.1 sec. So maybe it would be acceptable when we consider small scale normal usage in Non-Functional Requriement.

That's my feeling too, although actual data is needed to confirm. So, a good approach is to go with this simple approach (seems preferable from user's POV and the performance drawbacks are likely unnoticeable) and fallback on a more complicated alternative if the drawbacks are not as unnoticeable as we thought.

This situation is also relevant to the point about [_avoid premature optimizations_](https://nus-cs2103-ay2021s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-premature-optimizations) that will be covered in a future week.

**Comment by richardcom** [comment]
Hi @OhJunMing , just a gentle reminder to close the issue if the problem is resolved. : )
